### PR TITLE
feat: restructure scope3_methodology

### PR DESCRIPTION
**Pull request by EmmaLouise2018**
Purpose of this PR is to clean up the code, verify the logic and make it super clear what is happening (same units etc). This also abstract out some of the logic we were doing multiple times, and in general makes things cleaner with the yaml keys and models. This will pave the way to putting in a super simple API to interact with these models.

### CODE REVIEW

File: scope3_methodology/ad_tech_platform/model.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+                / self.bid_requests_processed_billion_per_month
+            )
+            log_result("external request ratio", f"{ratio * 100.0:.1f}%", depth)
+            return ratio
+        return self.get_decimal_fraction_of_bid_requests_processed_from_ad_tech_platforms()
+
+    def comp_bid_request_size_gb(self) -> float:
+        return self.bid_request_size_in_bytes / 1024 / 1024 / 1024
+
+    def comp_data_transfer_emissions_g_co2e_per_bid_request(self, depth: int) -> float:
```
Comment (EmmaLouise2018) [inline scope3_methodology/ad_tech_platform/model.py:142]: Have some questions about this function. Existing code is https://github.com/scope3data/methodology/blob/main/scope3_methodology/adtech_model.py#L125

**NOTE** This old function used to return `mt` if we had "data transfer emissions mt per month" and `g` if we had to use other facts to compute this number. Moved this to always return `g`. 

Why aren't we returning this divided by `get_bid_requests_processed_per_month()` ? Right now, this doesn't seem like its actually per bid request but the total emissions from data transfer

File: scope3_methodology/ad_tech_platform/model.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        return cookie_syncs_processed
+
+    def comp_water_usage_to_emissions_ratio_h2o_m_3_per_g_co2e(self, depth: int) -> float:
+        water_m3_per_emissions = (
+            self.datacenter_water_intensity_h2o_m_3_per_mwh / self.server_emissions_g_per_kwh / 1000
+        )
+        log_result("h2o m^3 per g co2e emissions", water_m3_per_emissions, depth)
+        return water_m3_per_emissions
+
+    def comp_primary_emissions_g_co2e_per_cookie_sync(
```
Comment (EmmaLouise2018) [inline scope3_methodology/ad_tech_platform/model.py:234]: It's not clear to me why are we multiplying the cookie sync emissions by water usage? This doesn't then return emissions by cookie sync. 

This function is currently `get_primary_emissions_per_cookie_sync` I renamed it to included `g_co2e` so we can be clear what we are computing, but in actuality this returns `water usage per cookie sync` Seems like this should be it's own function. https://github.com/scope3data/methodology/blob/main/scope3_methodology/adtech_model.py#L254 

i.e. 
```
    def comp_primary_emissions_g_co2e_per_cookie_sync(
        self,
        depth: int,
    ) -> float:
        primary_emissions_per_cookie_sync = (
            self.get_server_allocation_decimal_fraction()
            * self.get_server_emissions_g_co2e_per_month()
            * self.get_servers_processing_cookie_syncs_decimal_fraction()
        ) / self.comp_cookie_syncs_processed_per_month(depth)

        log_result("primary emissions g per cookie sync", primary_emissions_per_cookie_sync, depth)
       return primary_emissions_per_cookie_sync

   def comp_water_usage_per_cookie_sync(
        self,
        depth: int,
    ) -> float:
        water_usage_to_emissions_ratio = (
            self.comp_water_usage_to_emissions_ratio_h2o_m_3_per_g_co2e(depth - 1)
        )
        primary_water_usage = comp_primary_emissions_g_co2e_per_cookie_sync() * water_usage_to_emissions_ratio
        log_result("primary water usage m^3 per cookie sync", primary_water_usage, depth)
        return primary_water_usage
```

File: scope3_methodology/ad_tech_platform/model.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+            depth - 1
+        )
+        primary_emissions_g_co2e_per_bid_request = (
+            self.corporate_emissions_g_co2e_per_bid_request
+            + data_transfer_emissions_g_co2e_per_bid_request
+            + server_emissions_g_co2e_per_bid_request
+        )
+        log_result(
+            "primary emissions g co2e per bid request",
+            f"{primary_emissions_g_co2e_per_bid_request:.6f}",
```
Comment (EmmaLouise2018) [inline scope3_methodology/ad_tech_platform/model.py:197]: Updated this to be `g` not `mt` which is is currently https://github.com/scope3data/methodology/blob/main/scope3_methodology/adtech_model.py#L181

This was wrong in the existing code
```
primary_emissions_per_bid_request = (
        corporate_emissions_per_bid_request                   -> g 
        + data_transfer_emissions_per_bid_request         -> mt or g 
        + server_emissions_per_bid_request                     -> g 
    )
```

making sure we are in the same units

File: scope3_methodology/ad_tech_platform/model.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+        bid_request_distribution_rate: float,
+    ) -> None:
+        self.partner = partner
+        self.bid_request_distribution_rate = bid_request_distribution_rate
+
+
+class AdTechPlatform:
+    """Raw emissions information about an ATP and methodology of how to calculate emissions"""
+
+    def __init__(
```
Comment (MikeFreyberger) [inline scope3_methodology/ad_tech_platform/model.py:45]: I'd consider making this a dataclass to get the standard __init__ method for free: https://docs.python.org/3/library/dataclasses.html

File: scope3_methodology/ad_tech_platform/model.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+        )
+        self.cookie_syncs_processed_billion_per_month = cookie_syncs_processed_billion_per_month
+
+    def get_bid_requests_processed_per_month(self) -> float:
+        return self.bid_requests_processed_billion_per_month * 1000000000
+
+    def get_server_emissions_g_co2e_per_month(self) -> float:
+        return self.server_emissions_mt_co2e_per_month * 1000000
+
+    def get_servers_processing_bid_requests_pct(self) -> float:
```
Comment (bokelley) [inline scope3_methodology/ad_tech_platform/model.py:133]: nit but what do you think about reserving `pct` for when it's based on 100, and `rate` for when it's based on 1?

File: scope3_methodology/ad_tech_platform/model.py [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+
+    def get_servers_processing_cookie_syncs_pct(self) -> float:
+        return self.servers_processing_cookie_syncs_percentage / 100.0
+
+    def get_cookie_syncs_processed_per_month(self) -> float | None:
+        if self.cookie_syncs_processed_billion_per_month:
+            return self.cookie_syncs_processed_billion_per_month * 1000000000
+        return None
+
+    def get_corporate_allocation_decimal_fraction(self) -> float:
```
Comment (bokelley) [inline scope3_methodology/ad_tech_platform/model.py:144]: same point - rate?

File: scope3_methodology/ad_tech_platform/model.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+    def get_servers_processing_cookie_syncs_decimal_fraction(self) -> float:
+        return self.servers_processing_cookie_syncs_percentage / 100.0
+
+    def get_servers_processing_bid_requests_decimal_fraction(self) -> float:
+        return self.servers_processing_bid_requests_percentage / 100.0
+
+    def get_bid_request_rejection_decimal_fraction(self) -> float:
+        return self.bid_request_rejection_percentage / 100.0
+
+    def get_decimal_fraction_of_bid_requests_processed_from_ad_tech_platforms(self) -> float:
```
Comment (bokelley) [inline scope3_methodology/ad_tech_platform/model.py:156]: naming convention - unit first or last?

File: scope3_methodology/ad_tech_platform/model.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+        return self.percentage_of_bid_requests_processed_from_ad_tech_platforms / 100.0
+
+    def get_data_transfer_emissions_g_co2e_per_month(self) -> float | None:
+        if self.data_transfer_emissions_mt_co2e_per_month:
+            return self.data_transfer_emissions_mt_co2e_per_month * 1000000
```
Comment (bokelley) [inline scope3_methodology/ad_tech_platform/model.py:161]: worth throwing in a constant G_PER_MT?

File: scope3_methodology/ad_tech_platform/model.py (hunk 1 above)
Comment (bokelley) [reply scope3_methodology/ad_tech_platform/model.py:183]: As far as I can tell you are right - looks like a bug in my code

File: scope3_methodology/ad_tech_platform/model.py [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+        return None
+
+    def get_server_allocation_decimal_fraction(self) -> float:
+        return self.server_allocation_percentage / 100.0
+
+    def comp_external_request_ratio(self, depth: int) -> float:
```
Comment (bokelley) [inline scope3_methodology/ad_tech_platform/model.py:167]:
```suggestion
    # We assume that server-to-server requests (generally from ad tech platforms) incur core networking costs that are not accounted for in client device data transfer emissions. Therefore we need to figure out what portion of inbound requests are S2S.
    def comp_external_request_ratio(self, depth: int) -> float:
```

We probably should add a comment here with the assumption we're making and also rename something that is more like server-to-server requests. Unclear if we should be returning the ratio or just the number of S2S requests?

File: scope3_methodology/ad_tech_platform/model.py [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+            depth,
+        )
+        return data_transfer_emissions_g_co2e_per_bid_request
+
+    def comp_server_emissions_g_co2e_per_bid_request(
+        self,
+        depth: int,
+    ) -> float:
+        server_emissions_g_co2e_per_bid_request = (
+            self.get_server_allocation_decimal_fraction()
```
Comment (bokelley) [inline scope3_methodology/ad_tech_platform/model.py:208]: just reading code - is this allocation of parent org servers? Might be worth making that more clear

File: scope3_methodology/ad_tech_platform/model.py (hunk 2 above)
Comment (bokelley) [reply scope3_methodology/ad_tech_platform/model.py:283]: I'm sure it's copy & paste error on my part

File: scope3_methodology/ad_tech_platform/model.py [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+                f"{secondary_emissions_per_cookie_sync:.6f}",
+                depth,
+            )
+        return secondary_emissions_per_cookie_sync
+
+    def comp_corporate_emissions_g_co2e_per_bid_request(
+        self, corporate_emissions_g_co2e: float, depth: int
+    ) -> float:
+        emissions = (
+            self.get_corporate_allocation_decimal_fraction()
```
Comment (bokelley) [inline scope3_methodology/ad_tech_platform/model.py:320]: prob rename to be clear allocation of what?

File: scope3_methodology/ad_tech_platform/model.py [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+        self,
+        name: str,
+        identifier: str,
+        compute_secondary: bool,
+        depth: int,
+        corporate_emissions_g: float | None = None,
+        corporate_emissions_g_per_imp: float | None = None,
+    ):
+        # Compute corporate emissions
+        corporate_emissions_per_bid_request = corporate_emissions_g_per_imp
```
Comment (bokelley) [inline scope3_methodology/ad_tech_platform/model.py:346]: why is this per imp?

File: scope3_methodology/ad_tech_platform/yaml_keys.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python
+""" ATP Model Yaml Keys """
+from dataclasses import dataclass
+
+
+@dataclass
+class AdTechPlatformKeys:
```
Comment (bokelley) [inline scope3_methodology/ad_tech_platform/yaml_keys.py:7]: you are also allowed to just switch the yaml format to be snake case if that is easier & the team's preference!

File: scope3_methodology/compute_defaults.py [hunk 14]
```diff
@@ -3,11 +3,14 @@
...
-from adtech_model import get_ad_tech_model_keys
-from corporate import get_corporate_keys
-from utils import get_all_facts
+from ad_tech_platform.yaml_keys import AdTechPlatformKeys
+from corporate.yaml_keys import CorporateKeys
+from publisher.yaml_keys import PublisherPropertyKeys
+from utils.utils import get_all_facts
 from yaml.loader import SafeLoader
 
+# TODO make sure this is proucing defautls in the units we want now things have been updated
```
Comment (bokelley) [inline scope3_methodology/compute_defaults.py:12]: sp

File: scope3_methodology/corporate/model.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+            if commuting_emissions_mt_co2e_per_employee_per_month
+            else defaults.commuting_emissions_mt_per_employee_per_month
+        )
+        self.overhead_emissions_mt_co2e_per_employee_per_month = (
+            overhead_emissions_mt_co2e_per_employee_per_month
+            if overhead_emissions_mt_co2e_per_employee_per_month
+            else defaults.overhead_emissions_mt_per_employee_per_month
+        )
+
+    def validate(self) -> None:
```
Comment (bokelley) [inline scope3_methodology/corporate/model.py:51]: So if we don't have this data does this mean we fall over? I think we need to update this to be able to fall back to an average?

**Review by bokelley — COMMENTED** [review]
didn't get through all the files but submitting so you can absorb

File: scope3_methodology/ad_tech_platform/model.py (hunk 12 above)
Comment (EmmaLouise2018) [reply scope3_methodology/ad_tech_platform/model.py:346]: copy/paste error

File: scope3_methodology/ad_tech_platform/model.py (hunk 5 above)
Comment (EmmaLouise2018) [reply scope3_methodology/ad_tech_platform/model.py:133]: 👍

File: scope3_methodology/ad_tech_platform/yaml_keys.py (hunk 13 above)
Comment (EmmaLouise2018) [reply scope3_methodology/ad_tech_platform/yaml_keys.py:7]: Yeah, I'll switch that now. Was also going to put up a follow up PR that rearranges the yaml slightly as in https://www.notion.so/scope3/Public-Methodology-Setup-0ee430303d9144e99f942f7f019d3404#3af1bf647df448d0b4095c0a9e40e8f9

File: scope3_methodology/corporate/model.py (hunk 15 above)
Comment (EmmaLouise2018) [reply scope3_methodology/corporate/model.py:51]: Right now, yes we require employees or to know mt per month or we can't compute. You wrote up https://github.com/scope3data/methodology/issues/11 and should add in a way to fallback.

File: scope3_methodology/model_corporate_emissions.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+    args = parser.parse_args()
+    if args.verbose:
+        logging.basicConfig(level=logging.INFO)
+
+    # Load facts about the company
+    stream = open(args.companyFile[0], "r")
+    document = yaml.load(stream, Loader=SafeLoader)
+    if "company" not in document:
+        raise Exception("No 'company' field found in company file")
+    if "products" not in document["company"]:
```
Comment (bokelley) [inline scope3_methodology/model_corporate_emissions.py:40]: why does this need `products`?

**Comment by bokelley** [comment]
```
% ./scope3_methodology/fact_finder.py
Traceback (most recent call last):
  File "/Users/bokelley/src/methodology/./scope3_methodology/fact_finder.py", line 4, in <module>
    from utils import get_all_facts
ImportError: cannot import name 'get_all_facts' from 'utils' (/Users/bokelley/src/methodology/scope3_methodology/utils/__init__.py)

```

File: scope3_methodology/publisher/model.py [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+
+    def get_impressions(self):
+        return self.impressions
+
+    def set_corporate_emissions_g_co2e_per_impression(
+        self, emissions_g: float, emissions_g_per_imp: float
+    ) -> None:
+        if emissions_g:
+            self.corporate_emissions_per_impression = round(
+                emissions_g * self.ad_revenue_allocation_pct / self.impressions, 6
```
Comment (bokelley) [inline scope3_methodology/publisher/model.py:28]: I don't think 6 decimals is enough - goes to scientific notation for me

File: data/companies/theguardian/data.yaml [hunk 18]
```diff
@@ -56,3 +56,6 @@ company:
           commuting_emissions_mt_per_employee_per_month: 0.040
           travel_emissions_mt_per_employee_per_month: 0.137
           revenue_allocation_to_ads_pct: 33.3
+      -
+          corporate_emissions_g_per_impression: 0.000001778
```
Comment (EmmaLouise2018) [inline data/companies/theguardian/data.yaml:60]: When I run this locally I get `corporate_emissions_per_impression: 1.777946` ? 

```
➜ [Sep 20, 202210:33AM] ~/scope3/methodology git:(scop-211) python3 scope3_methodology/model_corporate_emissions.py publisher data/companies/theguardian/data.yaml 
corporate_emissions_mt_co2e_per_month: 794.5304000000001
➜ [Sep 20, 202210:34AM] ~/scope3/methodology git:(scop-211) ✗ python3 scope3_methodology/model_publisher_emissions.py data/companies/theguardian/data.yaml --corporateEmissionsG 794530400
properties:
- !!python/object:publisher.model.ModledProperty
  ad_revenue_allocation_pct: 23.309999999999995
  client_device_emissions_g_co2e_per_imp: 0.07246991860308909
  corporate_emissions_per_impression: 1.777946
  data_transfer_electricity_kwh: 1.0112473060344829e-05
  identifier: theguardian.com
  impressions: 10416800000.0
  page_load_electricity_kwh: 0.00012434006609195403

```

File: scope3_methodology/publisher/model.py [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+            )
+        else:
+            self.corporate_emissions_per_impression = round(emissions_g_per_imp, 6)
```
[the remark this answers is not in the document]
Comment (EmmaLouise2018) [reply scope3_methodology/publisher/model.py]: fixed

File: data/companies/theguardian/data.yaml (hunk 18 above)
Comment (EmmaLouise2018) [reply data/companies/theguardian/data.yaml]: I think we should stick to grams across the board, right now only corproate emissions is in mt. Everything else was in grams, so let's convert corporate to MT as well. Will make things easier to keep all within the same units. 

Nice call out about ad revenue allocation, found the bug. Was 
```self.corporate_emissions_g_co2e_per_impression = (
                emissions_g * self.ad_revenue_allocation_pct  / self.impressions
            )
```
should be 
```
self.corporate_emissions_g_co2e_per_impression = (
                emissions_g * self.get_ad_revenue_allocation_rate() / self.impressions
            )
```

Comment (EmmaLouise2018) [reply data/companies/theguardian/data.yaml]: That being said, we do get a lot of input units in mt. So only going to change the outputs to `g` The inputs will stay in `mt` unless we also want to change that ?

Comment (bokelley) [reply data/companies/theguardian/data.yaml]: agree - mt in, g out

File: scope3_methodology/corporate/yaml_keys.py [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+    commuting_emissions_mt_per_employee_per_month: str = (
+        "commuting_emissions_mt_per_employee_per_month"
+    )
+    datacenter_emissions_mt_per_employee_per_month: str = (
+        "datacenter_emissions_mt_per_employee_per_month"
+    )
+
+    def get_default_keys(self) -> set[str]:
+        return {
+            self.corporate_emissions_mt_per_month,
```
Comment (MikeFreyberger) [inline scope3_methodology/corporate/yaml_keys.py:24]: this key isn't being used in the __init__ method of `corporate_defaults`. is that intentional?

File: scope3_methodology/ad_tech_platform/helpers.py [hunk 21]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env python
+""" ATP Model Helpers """
+from utils.utils import log_step
+
+
+def get_product_info(key: str, default: float | None, product: dict[str, str], depth: int):
```
Comment (asweeney86) [inline scope3_methodology/ad_tech_platform/helpers.py:6]: Missing return type

File: scope3_methodology/model_ad_tech_platform.py [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+    distribution_partners: list[DistributionPartner],
+    corporate_emissions_g: float | None = None,
+    corporate_emissions_g_per_bid_request: float | None = None,
+    compute_secondary: bool = False,
+) -> ModeledAdTechPlatform:
+    # Validate product
+    if "name" not in product:
+        raise Exception("No 'name' field found in a product")
+
+    name = str(product["name"])
```
Comment (asweeney86) [inline scope3_methodology/model_ad_tech_platform.py:76]: Shouldn't need a cast here

File: scope3_methodology/publisher/yaml_keys.py [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+        if environment == "computer":
+            return self.computer_active_electricity_use_watts
+        elif environment == "mobile":
+            return self.mobile_active_electricity_use_watts
+        elif environment == "tv":
+            return self.tv_active_electricity_use_watts
+        else:
+            raise Exception(f"unsupported environment: {environment}")
+
+    def get_environment_idle_electricity_use_watts(self, environment: str):
```
Comment (asweeney86) [inline scope3_methodology/publisher/yaml_keys.py:40]: enum?

File: scope3_methodology/corporate/model.py [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+    corporate_emissions_mt_co2e_per_month: Optional[float] = field(
+        default=None, metadata={"default_eligible": False}
+    )
+    number_of_employees: Optional[int] = field(default=None, metadata={"default_eligible": False})
+    defaults: Optional["CorporateEmissions"] = field(
+        default=None, metadata={"default_eligible": False}
+    )
+
+    @classmethod
+    def default_fields(cls):
```
Comment (asweeney86) [inline scope3_methodology/corporate/model.py:39]: Create a base class so you don't need to duplicate this logic in each model.py?

**Review by asweeney86 — APPROVED** [review]

**Pull request merged**